internal/log/log.go

```diff
@@ -0,0 +1,151 @@
+// Package log provides logging utilities for the tracer.
+package log
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/version"
+)
+
+// Level specifies the logging level that the log package prints at.
+type Level int
+
+const (
+	// LevelDebug represents debug level messages.
+	LevelDebug Level = iota
+	// LevelWarn represents warning and errors.
+	LevelWarn
+)
+
+var prefixMsg = fmt.Sprintf("Datadog Tracer %s", version.Tag)
+
+var (
+	mu     sync.RWMutex   // guards below fields
+	level                 = LevelWarn
+	logger ddtrace.Logger = &defaultLogger{l: log.New(os.Stderr, "", log.LstdFlags)}
+)
+
+// UseLogger sets l as the active logger.
+func UseLogger(l ddtrace.Logger) {
+	mu.Lock()
+	defer mu.Unlock()
+	logger = l
+}
+
+// SetLevel sets the given lvl for logging.
+func SetLevel(lvl Level) {
+	mu.Lock()
+	defer mu.Unlock()
+	level = lvl
+}
+
+// Debug prints the given message if the level is LevelDebug.
+func Debug(fmt string, a ...interface{}) {
+	mu.RLock()
+	lvl := level
+	mu.RUnlock()
+	if lvl != LevelDebug {
+		return
+	}
+	printMsg("DEBUG", fmt, a...)
+}
+
+// Warn prints a warning message.
+func Warn(fmt string, a ...interface{}) {
+	printMsg("WARN", fmt, a...)
+}
+
+var (
+	errmu   sync.RWMutex                // guards below fields
+	erragg  = map[string]*errorReport{} // aggregated errors
+	errrate time.Duration               // the rate at which errors are reported
+	erron   bool                        // true if errors are being aggregated
+)
+
+func init() {
+	errrate = time.Minute
+	if v, ok := os.LookupEnv("DD_LOGGING_RATE"); ok {
+		if sec, err := strconv.Atoi(v); err != nil {
+			Warn("Invalid value for DD_LOGGING_RATE: %v", err)
+		} else {
+			errrate = time.Duration(sec) * time.Second
+		}
+	}
+}
+
+type errorReport struct {
+	err   error
+	count uint64
+}
+
+// Error aggregates errors under the given key. The aggregated errors are printed
+// once a minute or once every DD_LOGGING_RATE number of seconds.
+func Error(key, format string, a ...interface{}) {
```

Comment:
Could it automagically "key" on the caller function/line, or is that too expensive?

Reply:
This is the one part I'd like recommendations or alternatives on. I'm not entirely happy with the signature of this function and `key` is very ambiguous to API users, even if it's just an internal package.

I'm really fond of the idea of using the caller function / line. I think that's a feature coming in go1.13 and would be a fantastic way to do this. I'll try experimenting with using https://godoc.org/golang.org/x/xerrors#Errorf which is a preview of that. IIRC, it automatically includes caller information into the error.

Reply:
I've tried this.

It seems that even the `xerrors` package uses `runtime.Callers` under the hood, so we might as well use that directly. The performance difference between an `Error` caller without `runtime.Callers` and with is quite big:
```
benchmark            old ns/op     new ns/op     delta
BenchmarkError-4     20.7          646           +3020.77%
```
It is definitely not worth the cost of simplifying the API. I guess all in all, the uglier method (having the `key` argument) is best to ensure this won't be problematic. These error calls are quite frequent in some cases.

Reply:
Another idea would be to use `format` as the key. I think that most of the time it should be constant, unless someone commits to some wizardry which I think is safe to not account for.

Reply:
Ok, removing `key` and using `format` seems much nicer and closer to what we'd want the API to look like. I propose we go with that.